File: scripts/smoke-test.mjs
import { startMockBackend } from "./mock-backend.mjs";

const CLI_ARGS = new Set(process.argv.slice(2));
const OFFLINE_MODE = CLI_ARGS.has("--offline") || process.env.SMOKE_OFFLINE === "1";
const BASE_URL = process.env.SMOKE_BASE_URL || (OFFLINE_MODE ? "http://localhost:18000" : "http://localhost:8000");
const TENANT_ID = process.env.SMOKE_TENANT_ID || "tenant-a";
const ENTITY_ID = process.env.SMOKE_ENTITY_ID || "smoke-user";

async function callApi(method, path, payload) {
  const response = await fetch(`${BASE_URL}${path}`, {
    method,
    headers: { "Content-Type": "application/json" },
    body: payload ? JSON.stringify(payload) : undefined,
  });

  const text = await response.text();
  const body = text ? JSON.parse(text) : {};
  if (!response.ok) {
    throw new Error(`HTTP ${response.status} ${method} ${path}: ${text}`);
  }
  return body;
}

async function waitForHealth(maxWaitMs = 120000) {
  const start = Date.now();
  while (Date.now() - start < maxWaitMs) {
    try {
      const health = await callApi("GET", "/health");
      if (health.status === "ok") {
        console.log("[ok] backend health endpoint reachable");
        return;
      }
    } catch {
      // ignore until timeout
    }
    await new Promise((resolve) => setTimeout(resolve, 2000));
  }
  throw new Error("Backend health check timed out");
}

function assert(condition, message) {
  if (!condition) {
    throw new Error(message);
  }
}

async function main() {
  let mock;
  if (OFFLINE_MODE) {
    mock = await startMockBackend({ port: 18000 });
    console.log("[ok] started in-process mock backend on port 18000");
  }

  console.log(`Smoke test target: ${BASE_URL}`);
  try {
    await waitForHealth();

  const simulation = await callApi("POST", "/pipeline/simulate-attack-chain", {
    provider: "EntraID",
    tenantId: TENANT_ID,
    entityId: ENTITY_ID,
  });

  const incidents = simulation.incidents || [];
  assert(incidents.length > 0, "No incidents generated from attack-chain simulation");
  console.log(`[ok] simulation generated ${incidents.length} incident(s)`);

  const incidentsPayload = await callApi("GET", `/incidents?tenantId=${encodeURIComponent(TENANT_ID)}`);
  const incidentItems = incidentsPayload.items || [];
  assert(incidentItems.length > 0, "No incidents found in incident store");
  console.log(`[ok] incident store contains ${incidentItems.length} item(s)`);

  const tasksPayload = await callApi("GET", `/soar/tasks?tenantId=${encodeURIComponent(TENANT_ID)}`);
  const tasks = tasksPayload.items || [];
  assert(tasks.length > 0, "No SOAR tasks found after incident creation");
  const task = tasks[0];
  const taskId = task.taskId;
  console.log(`[ok] operating on SOAR task ${taskId}`);

  await callApi("POST", `/soar/tasks/${encodeURIComponent(taskId)}/assign`, { analyst: "smoke-analyst" });
  await callApi("POST", `/soar/tasks/${encodeURIComponent(taskId)}/approve`, {
    actor: "smoke-lead",
    note: "smoke approval",
  });
  await callApi("POST", `/soar/tasks/${encodeURIComponent(taskId)}/execute`, { actor: "smoke-bot" });
  console.log("[ok] assign/approve/execute sequence completed");

  const auditPayload = await callApi(
    "GET",
    `/soar/audit?tenantId=${encodeURIComponent(TENANT_ID)}&taskId=${encodeURIComponent(taskId)}`
  );
  const auditItems = auditPayload.items || [];
  const actions = new Set(auditItems.map((row) => row.action));

  assert(actions.has("assigned"), "SOAR audit missing assigned action");
  assert(actions.has("approved"), "SOAR audit missing approved action");
  assert(actions.has("executed"), "SOAR audit missing executed action");
  console.log(`[ok] audit trail validated for task ${taskId}`);

    console.log("Smoke test completed successfully");
  } finally {
    if (mock) {
      await mock.close();
      console.log("[ok] mock backend stopped");
    }
  }
}

main().catch((error) => {
  console.error(`[fail] ${error.message}`);
  process.exit(1);
});

File: scripts/smoke_test.py
import json
import os
import sys
import time
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.parse import urlencode
from urllib.request import Request, urlopen


BASE_URL = os.getenv("SMOKE_BASE_URL", "http://localhost:8000")
TENANT_ID = os.getenv("SMOKE_TENANT_ID", "tenant-a")
ENTITY_ID = os.getenv("SMOKE_ENTITY_ID", "smoke-user")


def call_api(method: str, path: str, payload: dict[str, Any] | None = None) -> Any:
    data = None
    headers = {"Content-Type": "application/json"}
    if payload is not None:
        data = json.dumps(payload).encode("utf-8")

    request = Request(f"{BASE_URL}{path}", data=data, headers=headers, method=method)
    try:
        with urlopen(request, timeout=20) as response:
            body = response.read().decode("utf-8")
            if not body:
                return {}
            return json.loads(body)
    except HTTPError as err:
        body = err.read().decode("utf-8") if err.fp else ""
        raise RuntimeError(f"HTTP {err.code} {method} {path}: {body}") from err
    except URLError as err:
        raise RuntimeError(f"Network error {method} {path}: {err}") from err


def wait_for_health(max_wait_seconds: int = 120) -> None:
    start = time.time()
    while time.time() - start < max_wait_seconds:
        try:
            response = call_api("GET", "/health")
            if response.get("status") == "ok":
                print("[ok] backend health endpoint reachable")
                return
        except Exception:
            pass
        time.sleep(2)
    raise RuntimeError("Backend health check timed out")


def assert_condition(condition: bool, message: str) -> None:
    if not condition:
        raise RuntimeError(message)


def main() -> int:
    print(f"Smoke test target: {BASE_URL}")
    wait_for_health()

    simulation = call_api(
        "POST",
        "/pipeline/simulate-attack-chain",
        {"provider": "EntraID", "tenantId": TENANT_ID, "entityId": ENTITY_ID},
    )
    incidents = simulation.get("incidents", [])
    assert_condition(len(incidents) > 0, "No incidents generated from attack-chain simulation")
    print(f"[ok] simulation generated {len(incidents)} incident(s)")

    incidents_payload = call_api("GET", f"/incidents?{urlencode({'tenantId': TENANT_ID})}")
    incident_items = incidents_payload.get("items", [])
    assert_condition(len(incident_items) > 0, "No incidents found in incident store")
    print(f"[ok] incident store contains {len(incident_items)} item(s)")

    tasks_payload = call_api("GET", f"/soar/tasks?{urlencode({'tenantId': TENANT_ID})}")
    tasks = tasks_payload.get("items", [])
    assert_condition(len(tasks) > 0, "No SOAR tasks found after incident creation")
    task = tasks[0]
    task_id = task["taskId"]
    print(f"[ok] operating on SOAR task {task_id}")

    call_api("POST", f"/soar/tasks/{task_id}/assign", {"analyst": "smoke-analyst"})
    call_api("POST", f"/soar/tasks/{task_id}/approve", {"actor": "smoke-lead", "note": "smoke approval"})
    call_api("POST", f"/soar/tasks/{task_id}/execute", {"actor": "smoke-bot"})
    print("[ok] assign/approve/execute sequence completed")

    audit_payload = call_api("GET", f"/soar/audit?{urlencode({'tenantId': TENANT_ID, 'taskId': task_id})}")
    audit_items = audit_payload.get("items", [])
    actions = {item.get("action") for item in audit_items}

    assert_condition("assigned" in actions, "SOAR audit missing assigned action")
    assert_condition("approved" in actions, "SOAR audit missing approved action")
    assert_condition("executed" in actions, "SOAR audit missing executed action")
    print(f"[ok] audit trail validated for task {task_id}")

    print("Smoke test completed successfully")
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as exc:
        print(f"[fail] {exc}")
        raise SystemExit(1)
